File: backend/routers/activities.py
from fastapi import APIRouter, Depends, Body, HTTPException, UploadFile, File, Form
from typing import List, Dict, Optional
from pydantic import BaseModel
from auth.clerk import is_clerk_user
from entities.user import User
from gateways.activities import (
    ActivitiesGateway,
    ActivityEntryAlreadyExistsException,
    ActivityDoesNotExistException,
)
from entities.activity import Activity, ActivityEntry, ImageInfo
from gateways.aws.s3 import S3Gateway
import uuid
import os
from services.notification_manager import NotificationManager
from gateways.users import UsersGateway
from datetime import datetime, timedelta, UTC
from loguru import logger
import traceback
from entities.notification import Notification
from fastapi import Request
from controllers.plan_controller import PlanController
from analytics.posthog import posthog
import threading

router = APIRouter()

activities_gateway = ActivitiesGateway()
notification_manager = NotificationManager()
users_gateway = UsersGateway()
plan_controller = PlanController()


class ActivityEntryResponse(BaseModel):
    id: str
    activity_id: str
    quantity: int
    date: str


@router.get("/activities", response_model=List[Activity])
async def get_activities(user: User = Depends(is_clerk_user)):
    return activities_gateway.get_all_activities_by_user_id(user.id)


@router.get("/activity-entries", response_model=List[ActivityEntryResponse])
async def get_activity_entries(user: User = Depends(is_clerk_user)):
    entries = activities_gateway.get_all_activity_entries_by_user_id(user.id)
    return [
        ActivityEntryResponse(
            id=e.id, activity_id=e.activity_id, quantity=e.quantity, date=e.date
        )
        for e in entries
    ]


@router.post("/log-activity", response_model=ActivityEntryResponse)
async def log_activity(
    activity_id: str = Form(...),
    iso_date_string: str = Form(...),
    quantity: int = Form(...),
    photo: UploadFile = None,
    isPublic: bool = Form(False),
    description: str = Form(None),
    timezone: str = Form(...),
    user: User = Depends(is_clerk_user),
):
    try:
        activity_entry = ActivityEntry.new(
            user_id=user.id,
            activity_id=activity_id,
            quantity=quantity,
            date=iso_date_string,
            description=description,
            timezone=timezone,
        )

        entry = None
        try:
            entry = activities_gateway.create_activity_entry(activity_entry)
        except ActivityEntryAlreadyExistsException:
            entry = activities_gateway.get_activity_entry_by_activity_and_date(
                activity_id, iso_date_string
            )
            if entry:
                entry = activities_gateway.update_activity_entry(
                    entry.id,
                    {
                        "quantity": quantity + entry.quantity,
                        "description": description,
                    },
                )
        except ActivityDoesNotExistException:
            raise HTTPException(status_code=404, detail="Activity does not exist")

        # Handle photo upload if provided
        if photo:
            s3_gateway = S3Gateway()
            photo_id = str(uuid.uuid4())
            _, file_extension = os.path.splitext(photo.filename)
            s3_path = f"/users/{user.id}/activity_entries/{entry.id}/photos/{photo_id}{file_extension}"

            s3_gateway.upload(await photo.read(), s3_path)

            expiration = 604799  # 7 days (max for s3 presigned url)
            presigned_url = s3_gateway.generate_presigned_url(s3_path, expiration)

            image_expires_at = datetime.now() + timedelta(seconds=expiration)

            image_info = ImageInfo(
                s3_path=s3_path,
                url=presigned_url,
                expires_at=image_expires_at.isoformat(),
                created_at=datetime.now().isoformat(),
                is_public=isPublic,
            )

            entry = activities_gateway.update_activity_entry(
                entry.id, {"image": image_info.dict()}
            )

            # Create notifications for friends about the photo
            for friend_id in user.friend_ids:
                logger.info(f"Creating notification for friend '{friend_id}'")
                activity = activities_gateway.get_activity_by_id(activity_id)
                message = f"{user.username} logged {quantity} {activity.measure} of {activity.emoji} {activity.title} with a photo 📸!"
                await notification_manager.create_and_process_notification(
                    Notification.new(
                        user_id=friend_id,
                        message=message,
                        type="info",
                        related_data={
                            "picture": user.picture,
                            "name": user.name,
                            "username": user.username,
                        },
                    )
                )

        # process coach update
        if user.plan_type != "free" and len(user.plan_ids) > 0:
            plan = plan_controller.get_plan(user.plan_ids[0])
            if activity_id in plan.activity_ids:
                # Run state transition processing in a separate thread to avoid blocking
                logger.info(f"Processing plan state transition for plan {plan.id}")
                thread = threading.Thread(
                    target=plan_controller.process_plan_state_recalculation,
                    args=(user, plan, False),
                    daemon=False,
                )
                thread.start()


        users_gateway.update_fields(
            user.id, {"last_active_at": datetime.now(UTC).isoformat()}
        )

        posthog.capture(
            distinct_id=user.id,
            event="activity_entry_created",
            properties={
                "activity_entry_id": entry.id,
                "user_id": user.id,
            },
        )
        return entry

    except Exception as e:
        logger.error(f"Error logging activity: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to log activity")


@router.get("/recent-activities")
async def get_recent_activities(user: User = Depends(is_clerk_user)):
    recent_activities = activities_gateway.get_readable_recent_activity_entries(
        user.id, limit=5
    )
    return {"recent_activities": recent_activities}


@router.post("/upsert-activity")
async def upsert_activity(
    activity: dict = Body(...), user: User = Depends(is_clerk_user)
):
    activity_id = activity.get("id", None)
    existent_activity = activities_gateway.get_activity_by_id(activity_id)

    if existent_activity:
        if existent_activity.user_id != user.id:
            raise HTTPException(
                status_code=403, detail="Not authorized to update this activity"
            )
        try:
            updated_activity = activities_gateway.update_activity(Activity(**activity))
            return updated_activity
        except ActivityDoesNotExistException:
            pass

    new_activity = Activity.new(
        id=activity_id,
        user_id=user.id,
        title=activity["title"],
        measure=activity["measure"],
        emoji=activity["emoji"],
        privacy_settings=activity["privacy_settings"],
    )
    created_activity = activities_gateway.create_activity(new_activity)

    return created_activity


@router.put("/activity-entries/{activity_entry_id}")
async def update_activity_entry(
    activity_entry_id: str,
    update_data: dict = Body(...),
    user: User = Depends(is_clerk_user),
):
    try:
        # Get the existing entry to verify ownership
        existing_entry = activities_gateway.get_activity_entry_by_id(activity_entry_id)
        if not existing_entry:
            raise HTTPException(status_code=404, detail="Activity entry not found")

        if existing_entry.user_id != user.id:
            raise HTTPException(
                status_code=403, detail="Not authorized to update this entry"
            )

        # Update the entry
        updated_entry = activities_gateway.update_activity_entry(
            activity_entry_id,
            {
                "quantity": update_data.get("quantity", existing_entry.quantity),
                "date": update_data.get("date", existing_entry.date),
                "description": update_data.get(
                    "description", existing_entry.description
                ),
            },
        )

        return updated_entry
    except Exception as e:
        logger.error(
            f"Error updating activity entry: {str(e)}\n{traceback.format_exc()}"
        )
        raise HTTPException(status_code=500, detail="Failed to update activity entry")


@router.post("/activity-entries/{activity_entry_id}/reactions")
async def add_activity_reaction(
    activity_entry_id: str,
    emoji: str = Body(None),
    emojis: List[str] = Body(None),
    operation: str = Body(...),
    user: User = Depends(is_clerk_user),
):
    try:
        if operation == "add":
            # Handle both single emoji and multiple emojis
            emoji_list = emojis if emojis else [emoji] if emoji else []
            if not emoji_list:
                raise HTTPException(status_code=400, detail="No emojis provided")

            updated_entry = None
            for e in emoji_list:
                updated_entry = activities_gateway.add_reaction(
                    activity_entry_id=activity_entry_id, emoji=e, user=user
                )

            # Only send notification once with all emojis
            activity_entry = activities_gateway.get_activity_entry_by_id(
                activity_entry_id
            )

            # Create a message with all the emojis
            if len(emoji_list) <= 3:
                emoji_text = " ".join(emoji_list)
            else:
                emoji_text = (
                    " ".join(emoji_list[:3]) + f" and {len(emoji_list) - 3} more"
                )

            await notification_manager.create_and_process_notification(
                Notification.new(
                    user_id=activity_entry.user_id,
                    message=f"@{user.username} reacted to your activity with {emoji_text}",
                    type="info",
                    related_data={
                        "picture": user.picture,
                        "name": user.name,
                        "username": user.username,
                    },
                )
            )
            return {"message": "Reactions added successfully", "entry": updated_entry}
        elif operation == "remove":
            # Handle both single emoji and multiple emojis
            emoji_list = emojis if emojis else [emoji] if emoji else []
            if not emoji_list:
                raise HTTPException(status_code=400, detail="No emojis provided")

            updated_entry = None
            for e in emoji_list:
                updated_entry = activities_gateway.remove_reaction(
                    activity_entry_id=activity_entry_id, emoji=e, user=user
                )
            return {"message": "Reactions removed successfully", "entry": updated_entry}
        else:
            raise HTTPException(status_code=400, detail="Invalid operation")
    except Exception as e:
        logger.error(f"Error managing reactions: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to manage reactions")


@router.get("/activity-entries/{activity_entry_id}/reactions")
async def get_activity_reactions(
    activity_entry_id: str,
    user: User = Depends(is_clerk_user),
):
    try:
        entry = activities_gateway.get_activity_entry_by_id(activity_entry_id)
        if not entry:
            raise HTTPException(status_code=404, detail="Activity entry not found")

        # Convert user_ids to usernames for the frontend
        reactions_with_usernames = {}
        for emoji, usernames in entry.reactions.items():
            reactions_with_usernames[emoji] = usernames

        return {"reactions": reactions_with_usernames}
    except Exception as e:
        logger.error(f"Error getting reactions: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to get reactions")


@router.delete("/activities/{activity_id}")
async def delete_activity(
    activity_id: str,
    user: User = Depends(is_clerk_user),
):
    try:
        # Get the activity to verify ownership
        activity = activities_gateway.get_activity_by_id(activity_id)
        if not activity:
            raise HTTPException(status_code=404, detail="Activity not found")

        if activity.user_id != user.id:
            raise HTTPException(
                status_code=403, detail="Not authorized to delete this activity"
            )

        # Check if activity is used in any active plans
        if plan_controller.is_activity_in_any_active_plan(activity_id):
            raise HTTPException(
                status_code=400,
                detail="Please remove this activity from all active plans before deleting it.",
            )

        # Delete the activity
        activities_gateway.delete_activity(activity_id)
        return {"message": "Activity deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting activity: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to delete activity")


@router.delete("/activity-entries/{activity_entry_id}")
async def delete_activity_entry(
    activity_entry_id: str,
    user: User = Depends(is_clerk_user),
):
    try:
        # Get the activity to verify ownership
        activity_entry = activities_gateway.get_activity_entry_by_id(activity_entry_id)
        if not activity_entry:
            raise HTTPException(status_code=404, detail="Activity not found")

        if activity_entry.user_id != user.id:
            raise HTTPException(
                status_code=403, detail="Not authorized to delete this activity"
            )

        # Delete the activity
        activities_gateway.delete_activity_entry(activity_entry_id)
        return {"message": "Activity deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting activity: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to delete activity")


# Add new response model for comments
@router.post("/activity-entries/{activity_entry_id}/comments")
async def add_activity_comment(
    activity_entry_id: str,
    request: Request,
    user: User = Depends(is_clerk_user),
):
    try:
        body = await request.json()
        text = body["text"]
        updated_entry = activities_gateway.add_comment(
            activity_entry_id=activity_entry_id, text=text, user=user
        )

        # Get the activity entry owner
        activity_entry = activities_gateway.get_activity_entry_by_id(activity_entry_id)

        # Don't notify if commenting on own activity
        if activity_entry.user_id != user.id:
            # Create notification for the activity owner
            await notification_manager.create_and_process_notification(
                Notification.new(
                    user_id=activity_entry.user_id,
                    message=f"@{user.username} commented on your activity: \"{text[:30]}{'...' if len(text) > 30 else ''}\"",
                    type="info",
                    related_data={
                        "picture": user.picture,
                        "name": user.name,
                        "username": user.username,
                    },
                )
            )

        # Return the latest comment
        latest_comment = updated_entry.comments[-1]
        return {
            "id": latest_comment.id,
            "user_id": latest_comment.user_id,
            "username": latest_comment.username,
            "text": latest_comment.text,
            "created_at": latest_comment.created_at,
            "picture": latest_comment.picture,
        }
    except Exception as e:
        logger.error(f"Error adding comment: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to add comment")


@router.delete("/activity-entries/{activity_entry_id}/comments/{comment_id}")
async def remove_activity_comment(
    activity_entry_id: str,
    comment_id: str,
    user: User = Depends(is_clerk_user),
):
    try:
        updated_entry = activities_gateway.remove_comment(
            activity_entry_id=activity_entry_id, comment_id=comment_id, user=user
        )
        return {"message": "Comment removed successfully"}
    except Exception as e:
        logger.error(f"Error removing comment: {str(e)}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail="Failed to remove comment")

File: backend-node/src/routes/activities.ts
import { Router, Response } from 'express';
import multer from 'multer';
import { requireAuth, AuthenticatedRequest } from '@/middleware/auth';
import { logger } from '../utils/logger';
import { prisma } from '../utils/prisma';
import { notificationService } from '../services/notificationService';
import { s3Service } from '../services/s3Service';
import { ActivityVisibility } from '@prisma/client';
import { v4 as uuidv4 } from 'uuid';

const router = Router();
const upload = multer({ storage: multer.memoryStorage() });

// Get all activities for user
router.get('/activities', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const activities = await prisma.activity.findMany({
      where: {
        userId: req.user!.id,
        deletedAt: null,
      },
      orderBy: { createdAt: 'desc' },
    });

    res.json(activities);
  } catch (error) {
    logger.error('Error fetching activities:', error);
    res.status(500).json({ error: 'Failed to fetch activities' });
  }
});

// Get all activity entries for user
router.get('/activity-entries', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const entries = await prisma.activityEntry.findMany({
      where: {
        userId: req.user!.id,
        deletedAt: null,
      },
      orderBy: { createdAt: 'desc' },
    });

    const response = entries.map(entry => ({
      id: entry.id,
      activityId: entry.activityId,
      quantity: entry.quantity,
      date: entry.date,
    }));

    res.json(response);
  } catch (error) {
    logger.error('Error fetching activity entries:', error);
    res.status(500).json({ error: 'Failed to fetch activity entries' });
  }
});

// Log activity entry
router.post('/log-activity', requireAuth, upload.single('photo'), async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityId, iso_date_string, quantity, isPublic, description, timezone } = req.body;
    const photo = req.file;

    // Check if activity exists and belongs to user
    const activity = await prisma.activity.findFirst({
      where: {
        id: activityId,
        userId: req.user!.id,
        deletedAt: null,
      },
    });

    if (!activity) {
      return res.status(404).json({ error: 'Activity not found' });
    }

    // Check if entry already exists for this date
    const existingEntry = await prisma.activityEntry.findFirst({
      where: {
        activityId: activityId,
        userId: req.user!.id,
        date: iso_date_string,
        deletedAt: null,
      },
    });

    let entry;
    if (existingEntry) {
      // Update existing entry by adding quantity
      entry = await prisma.activityEntry.update({
        where: { id: existingEntry.id },
        data: {
          quantity: existingEntry.quantity + parseInt(quantity),
          description: description || existingEntry.description,
        },
      });
    } else {
      // Create new entry
      entry = await prisma.activityEntry.create({
        data: {
          activityId: activityId,
          userId: req.user!.id,
          quantity: parseInt(quantity),
          date: iso_date_string,
          description,
          timezone,
        },
      });
    }

    // Handle photo upload if provided
    if (photo) {
      try {
        const photoId = uuidv4();
        const fileExtension = photo.originalname ? photo.originalname.split('.').pop() : 'jpg';
        const s3Path = `/users/${req.user!.id}/activity_entries/${entry.id}/photos/${photoId}.${fileExtension}`;

        // Upload to S3
        await s3Service.upload(photo.buffer, s3Path, photo.mimetype);

        // Generate presigned URL with 7 days expiration
        const expirationSeconds = 7 * 24 * 60 * 60; // 7 days
        const presignedUrl = await s3Service.generatePresignedUrl(s3Path, expirationSeconds);
        const expiresAt = new Date(Date.now() + expirationSeconds * 1000);

        // Update entry with image information
        entry = await prisma.activityEntry.update({
          where: { id: entry.id },
          data: {
            imageS3Path: s3Path,
            imageUrl: presignedUrl,
            imageExpiresAt: expiresAt,
            imageCreatedAt: new Date(),
            imageIsPublic: isPublic === 'true' || isPublic === true,
          },
        });

        logger.info(`Photo uploaded successfully to S3: ${s3Path}`);

        // Create notifications for connected users about the photo
        const userWithConnections = await prisma.user.findUnique({
          where: { id: req.user!.id },
          include: { 
            connectionsFrom: {
              where: { status: 'ACCEPTED' },
              include: { to: true }
            },
            connectionsTo: {
              where: { status: 'ACCEPTED' },
              include: { from: true }
            }
          },
        });

        if (userWithConnections) {
          const connectedUsers = [
            ...userWithConnections.connectionsFrom.map(conn => conn.to),
            ...userWithConnections.connectionsTo.map(conn => conn.from)
          ];
          
          for (const connectedUser of connectedUsers) {
            const message = `${req.user!.username} logged ${quantity} ${activity.measure} of ${activity.emoji} ${activity.title} with a photo 📸!`;
            await notificationService.createAndProcessNotification({
              userId: connectedUser.id,
              message,
              type: 'INFO',
              relatedData: {
                picture: req.user!.picture,
                name: req.user!.name,
                username: req.user!.username,
              },
            });
          }
        }
      } catch (error) {
        logger.error('Error uploading photo to S3:', error);
        // Continue without photo - don't fail the entire activity logging
      }
    }

    // Update user's last active time
    await prisma.user.update({
      where: { id: req.user!.id },
      data: {
        lastActiveAt: new Date(),
      },
    });

    // TODO: Add PostHog analytics
    // TODO: Add plan state processing

    res.json({
      id: entry.id,
      activityId: entry.activityId,
      quantity: entry.quantity,
      date: entry.date,
    });
  } catch (error) {
    logger.error('Error logging activity:', error);
    res.status(500).json({ error: 'Failed to log activity' });
  }
});

// Get recent activities
router.get('/recent-activities', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const recentEntries = await prisma.activityEntry.findMany({
      where: {
        userId: req.user!.id,
        deletedAt: null,
      },
      include: {
        activity: true,
      },
      orderBy: { createdAt: 'desc' },
      take: 5,
    });

    const recentActivities = recentEntries.map(entry => ({
      id: entry.id,
      activity_title: entry.activity.title,
      activity_emoji: entry.activity.emoji,
      quantity: entry.quantity,
      measure: entry.activity.measure,
      date: entry.date,
      createdAt: entry.createdAt,
    }));

    res.json({ recent_activities: recentActivities });
  } catch (error) {
    logger.error('Error fetching recent activities:', error);
    res.status(500).json({ error: 'Failed to fetch recent activities' });
  }
});

// Upsert activity
router.post('/upsert-activity', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { id, title, measure, emoji, privacy_settings } = req.body;

    if (id) {
      // Check if activity exists and belongs to user
      const existingActivity = await prisma.activity.findFirst({
        where: {
          id,
          userId: req.user!.id,
          deletedAt: null,
        },
      });

      if (existingActivity) {
        const updatedActivity = await prisma.activity.update({
          where: { id },
          data: {
            title,
            measure,
            emoji,
            privacySettings: privacy_settings as ActivityVisibility,
          },
        });
        return res.json(updatedActivity);
      }
    }

    // Create new activity
    const newActivity = await prisma.activity.create({
      data: {
        id,
        userId: req.user!.id,
        title,
        measure,
        emoji,
        privacySettings: privacy_settings as ActivityVisibility,
      },
    });

    res.json(newActivity);
  } catch (error) {
    logger.error('Error upserting activity:', error);
    res.status(500).json({ error: 'Failed to upsert activity' });
  }
});

// Update activity entry
router.put('/activity-entries/:activityEntryId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId } = req.params;
    const { quantity, date, description } = req.body;

    // Verify ownership
    const existingEntry = await prisma.activityEntry.findFirst({
      where: {
        id: activityEntryId,
        userId: req.user!.id,
        deletedAt: null,
      },
    });

    if (!existingEntry) {
      return res.status(404).json({ error: 'Activity entry not found' });
    }

    const updatedEntry = await prisma.activityEntry.update({
      where: { id: activityEntryId },
      data: {
        quantity: quantity !== undefined ? quantity : existingEntry.quantity,
        date: date || existingEntry.date,
        description: description !== undefined ? description : existingEntry.description,
      },
    });

    res.json(updatedEntry);
  } catch (error) {
    logger.error('Error updating activity entry:', error);
    res.status(500).json({ error: 'Failed to update activity entry' });
  }
});

// Add reaction to activity entry
router.post('/activity-entries/:activityEntryId/reactions', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId } = req.params;
    const { emoji, emojis, operation } = req.body;

    const emojiList = emojis || (emoji ? [emoji] : []);
    
    if (!emojiList.length) {
      return res.status(400).json({ error: 'No emojis provided' });
    }

    if (operation === 'add') {
      for (const e of emojiList) {
        await prisma.reaction.upsert({
          where: {
            activityEntryId_userId_emoji: {
              activityEntryId,
              userId: req.user!.id,
              emoji: e,
            },
          },
          update: {},
          create: {
            activityEntryId,
            userId: req.user!.id,
            emoji: e,
          },
        });
      }

      // Get activity entry to notify owner
      const activityEntry = await prisma.activityEntry.findUnique({
        where: { id: activityEntryId },
        include: { user: true },
      });

      if (activityEntry && activityEntry.userId !== req.user!.id) {
        const emojiText = emojiList.length <= 3 
          ? emojiList.join(' ')
          : `${emojiList.slice(0, 3).join(' ')} and ${emojiList.length - 3} more`;

        await notificationService.createAndProcessNotification({
          userId: activityEntry.userId,
          message: `@${req.user!.username} reacted to your activity with ${emojiText}`,
          type: 'INFO',
          relatedData: {
            picture: req.user!.picture,
            name: req.user!.name,
            username: req.user!.username,
          },
        });
      }

      res.json({ message: 'Reactions added successfully' });
    } else if (operation === 'remove') {
      for (const e of emojiList) {
        await prisma.reaction.deleteMany({
          where: {
            activityEntryId,
            userId: req.user!.id,
            emoji: e,
          },
        });
      }

      res.json({ message: 'Reactions removed successfully' });
    } else {
      res.status(400).json({ error: 'Invalid operation' });
    }
  } catch (error) {
    logger.error('Error managing reactions:', error);
    res.status(500).json({ error: 'Failed to manage reactions' });
  }
});

// Get reactions for activity entry
router.get('/activity-entries/:activityEntryId/reactions', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId } = req.params;

    const reactions = await prisma.reaction.findMany({
      where: { activityEntryId },
      include: { user: { select: { username: true } } },
    });

    // Group reactions by emoji
    const reactionsGrouped: { [emoji: string]: string[] } = {};
    reactions.forEach(reaction => {
      if (!reactionsGrouped[reaction.emoji]) {
        reactionsGrouped[reaction.emoji] = [];
      }
      reactionsGrouped[reaction.emoji].push(reaction.user.username!);
    });

    res.json({ reactions: reactionsGrouped });
  } catch (error) {
    logger.error('Error getting reactions:', error);
    res.status(500).json({ error: 'Failed to get reactions' });
  }
});

// Delete activity
router.delete('/activities/:activityId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityId } = req.params;

    // Verify ownership
    const activity = await prisma.activity.findFirst({
      where: {
        id: activityId,
        userId: req.user!.id,
        deletedAt: null,
      },
    });

    if (!activity) {
      return res.status(404).json({ error: 'Activity not found' });
    }

    // TODO: Check if activity is used in any active plans

    // Soft delete
    await prisma.activity.update({
      where: { id: activityId },
      data: { deletedAt: new Date() },
    });

    res.json({ message: 'Activity deleted successfully' });
  } catch (error) {
    logger.error('Error deleting activity:', error);
    res.status(500).json({ error: 'Failed to delete activity' });
  }
});

// Delete activity entry
router.delete('/activity-entries/:activityEntryId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId } = req.params;

    // Verify ownership
    const activityEntry = await prisma.activityEntry.findFirst({
      where: {
        id: activityEntryId,
        userId: req.user!.id,
        deletedAt: null,
      },
    });

    if (!activityEntry) {
      return res.status(404).json({ error: 'Activity entry not found' });
    }

    // Soft delete
    await prisma.activityEntry.update({
      where: { id: activityEntryId },
      data: { deletedAt: new Date() },
    });

    res.json({ message: 'Activity entry deleted successfully' });
  } catch (error) {
    logger.error('Error deleting activity entry:', error);
    res.status(500).json({ error: 'Failed to delete activity entry' });
  }
});

// Add comment to activity entry
router.post('/activity-entries/:activityEntryId/comments', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId } = req.params;
    const { text } = req.body;

    const comment = await prisma.comment.create({
      data: {
        activityEntryId,
        userId: req.user!.id,
        username: req.user!.username!,
        text,
      },
    });

    // Get activity entry to notify owner
    const activityEntry = await prisma.activityEntry.findUnique({
      where: { id: activityEntryId },
    });

    if (activityEntry && activityEntry.userId !== req.user!.id) {
      const truncatedText = text.length > 30 ? `${text.slice(0, 30)}...` : text;
      await notificationService.createAndProcessNotification({
        userId: activityEntry.userId,
        message: `@${req.user!.username} commented on your activity: "${truncatedText}"`,
        type: 'INFO',
        relatedData: {
          picture: req.user!.picture,
          name: req.user!.name,
          username: req.user!.username,
        },
      });
    }

    res.json({
      id: comment.id,
      userId: comment.userId,
      username: comment.username,
      text: comment.text,
      createdAt: comment.createdAt,
      picture: req.user!.picture,
    });
  } catch (error) {
    logger.error('Error adding comment:', error);
    res.status(500).json({ error: 'Failed to add comment' });
  }
});

// Remove comment from activity entry
router.delete('/activity-entries/:activityEntryId/comments/:commentId', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { activityEntryId, commentId } = req.params;

    // Verify ownership
    const comment = await prisma.comment.findFirst({
      where: {
        id: commentId,
        activityEntryId,
        userId: req.user!.id,
      },
    });

    if (!comment) {
      return res.status(404).json({ error: 'Comment not found' });
    }

    await prisma.comment.delete({
      where: { id: commentId },
    });

    res.json({ message: 'Comment removed successfully' });
  } catch (error) {
    logger.error('Error removing comment:', error);
    res.status(500).json({ error: 'Failed to remove comment' });
  }
});

export const activitiesRouter = router;
export default activitiesRouter;